### PR TITLE
[22.05] Miscellaneous security patches/backports

### PR DESCRIPTION
**Pull request by Ma27**
###### Description of changes

While updating nixpkgs 22.05 in a project and running vulnix I realized that quite a bunch of CVE fixes weren't backported (or patched if a backport of a bump is not viable) so I decided to bite the bullet and do it in one go:

* `vim` (backport for #187868)
  * Fixes CVE-2022-2816, CVE-2022-2817, CVE-2022-2819, CVE-2022-2845, CVE-2022-2849, CVE-2022-2862, CVE-2022-2874, CVE-2022-2889.
* `go_1_17` (backport for #176951, #181284, #184636)
  * Fixes CVE-2022-1705, CVE-2022-32148, CVE-2022-30631, CVE-2022-30633, CVE-2022-28131, CVE-2022-30635, CVE-2022-30632, CVE-2022-30630, CVE-2022-1962
* `go_1_18` (backport of #176951, #181281, #184617)
  * Fixes CVE-2022-1705, CVE-2022-32148, CVE-2022-30631, CVE-2022-30633, CVE-2022-28131, CVE-2022-30635, CVE-2022-30632, CVE-2022-30630, CVE-2022-1962
* `rsync` (backport of #174894, #187542)
  * Fixes CVE-2022-29154
* `minio` (cannot backport further updates, so manually applied the patch)
  * Fixes CVE-2022-35919
* `libtiff` (cannot backport further updates, so manually applied the patch)
  * Fixes CVE-2022-2867, CVE-2022-2868, CVE-2022-2869
* `sqlite` (cannot backport further updates, so manually applied the patch)
  * Fixes CVE-2022-35737

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by winterqt** [comment]
Where did you get the SQLite patch? The only commit I can see that mentions the CVE is [this](https://www.sqlite.org/src/vinfo/aab790a16e1bdff7), which doesn't look like yours at all.

**Comment by Ma27** [comment]
@winterqt thanks! I somehow selected the wrong one in their history :facepalm: 

Sorry for that, will update now!

**Comment by risicle** [comment]
libtiff patch hasn't made it to unstable yet, looks like jumped past me on the nvd rss feed.

File: pkgs/applications/networking/sync/rsync/default.nix [hunk 1]
```diff
@@ -16,32 +15,27 @@
 , xxHash
 , enableZstd ? true
 , zstd
-, enableCopyDevicesPatch ? false
```
Comment (risicle) [inline pkgs/applications/networking/sync/rsync/default.nix:19]: Removing this option changes the calling convention of this package, which is not something we should do in stable.

Comment (winterqt) [reply pkgs/applications/networking/sync/rsync/default.nix:19]: The option that the patch adds is included in this version, so we should keep the option but make it a no-op. (Also maybe warn that it's superfluous?)

Comment (Ma27) [reply pkgs/applications/networking/sync/rsync/default.nix:19]: Done as suggested by @winterqt .
For further context see https://github.com/WayneD/rsync/blob/v3.2.5/NEWS.md#enhancements-1.

File: pkgs/applications/networking/sync/rsync/default.nix [hunk 2]
```diff
@@ -16,32 +15,33 @@
...
 , nixosTests
+, enableCopyDevicesPatch ? false
 }:
 
 let
-  base = import ./base.nix { inherit lib fetchurl fetchpatch; };
+  mkDerivation = if enableCopyDevicesPatch
+    then lib.warn "enableCopyDevicesPatch in pkgs.rsync is not needed anymore" stdenv.mkDerivation
```
Comment (winterqt) [inline pkgs/applications/networking/sync/rsync/default.nix:24]: Prefer `lib.warnIf` here, which will make this entire expression simpler.

**Review by winterqt — CHANGES_REQUESTED** [review]

**Review by fpletz — APPROVED** [review]

**Review by risicle — CHANGES_REQUESTED** [review]
The libtiff commit does *not* address all three vulnerabilities. It only fixes CVE-2022-2869.

Note the MR linked from the issues (https://gitlab.com/libtiff/libtiff/-/merge_requests/294) consists of *three* commits. If you were targeting the combined result, that would have been the merge commit https://gitlab.com/libtiff/libtiff/-/commit/d46af879bc4e789f95a7828fa566f61700954393

Source: personally trying the poc for each of the issues with the patched and unpatched libtiff.

**Comment by risicle** [comment]
It's a lot harder to do a proper review when things are all bundled together like this.

**Comment by Ma27** [comment]
Oof, yeah seems like I was really good at getting trolled by different UIs today :D 

Push will follow in a minute :)

**Comment by risicle** [comment]
`libtiff` patch now resolves all 4 test cases for the 3 CVEs.

**Comment by risicle** [comment]
`libtiff.tests` builds for me, macos 10.15, nixos x86_64

**Comment by Ma27** [comment]
ping @winterqt @risicle is there anything else tbd to get this merged? :)

**Review by risicle — APPROVED** [review]

**Comment by risicle** [comment]
..but I haven't looked at anything other than the libtiff changes

File: pkgs/servers/minio/default.nix [hunk 3]
```diff
@@ -26,6 +26,13 @@ buildGoModule rec {
 
   vendorSha256 = "sha256-u36oHqIxMD3HRio9A3tUt6FO1DtAcQDiC/O54ByeNyg=";
 
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/minio/minio/commit/bc72e4226e669d98c8e0f3eccc9297be9251c692.patch";
+      sha256 = "sha256-420jff2B1eZArUs8NncWX6dQMMTGAfF1nNjzbwo6AUI=";
+    })
```
Comment (winterqt) [inline pkgs/servers/minio/default.nix:33]: We should set `name` to the CVE ID here.

**Review by winterqt — CHANGES_REQUESTED** [review]
Do you mind collapsing the Go bumps into one commit per version? For example, take [go_1_17: 1.17.10 -> 1.17.11](https://github.com/NixOS/nixpkgs/pull/187893/commits/614665a7eec607164b9a8dddfbefee91a644dc6b), [go_1_17: 1.17.11 -> 1.17.12](https://github.com/NixOS/nixpkgs/pull/187893/commits/2c3222bcc088741c2f78bd6998e21cb540d80b04), and [go_1_17: 1.17.12 -> 1.17.13](https://github.com/NixOS/nixpkgs/pull/187893/commits/245c28bdc7e24281b1398404ff7ec08d33d1f118), and make one commit titled `go_1_17: 1.17.10 -> 1.17.13`. I think it's a lot of extra noise for no good reason, though I'm open to discussing this if you think I'm being overly nitpicky.

(I'd suggest the same for rsync, but keeping those separated might be good since it's \<bump>, \<refactor>, \<bump>. That first commit can be removed, though it's probably not worth the effort.)

Also, please cherry-pick bf0aa680881de6d6523b73b35b3f66977ded3c93. (Sandro's refactor re-added this when it was removed in the previous bump, presumably because it was older than that bump.)

**Comment by Ma27** [comment]
@winterqt fixed the patch name and squashed to go commits together.

I agree that in case of rsync this might be confusing considering the refactor, so I didn't do it.

**Review by winterqt — APPROVED** [review]

**Comment by winterqt** [comment]
> (cherry picked from commit https://github.com/NixOS/nixpkgs/commit/47016de7effddab56c374598de8de7c704543c84)
> (cherry picked from commit https://github.com/NixOS/nixpkgs/commit/6b6a73f53e600267d4016775270a2e770178d4e6)
> (cherry picked from commit https://github.com/NixOS/nixpkgs/commit/8fb01fab4f38dbcf6e5e55a786aef0d77fe28865)

Not sure how I feel about these in the Go bumps' descriptions (ideally it would just be the last commit, we don't care about the intermediate state), but I thought asking you to change it would be way too nit-picky of me since it doesn't matter in any actual way 😅

Curious: did you add that manually, or did Git do it? If it was a conscious choice, why did you think that was preferable to having one cherry-pick note, or even none at all?

**Comment by Ma27** [comment]
> Curious: did you add that manually, or did Git do it? If it was a conscious choice, why did you think that was preferable to having one cherry-pick note, or even none at all?

I squashed the go bumps into a single one and figured that it's still reasonable to keep the references to the original commit**s**, so I left all of them on purpose.

May not be the prettiest solution, but is still rather useful IMHO.

**Comment by vcunat** [comment]
`sqlite`: we have this top-level attribute, both in nixpkgs master and 22.05:
```nix
{
  sqlite-replication = sqlite.overrideAttrs (oldAttrs: rec {
    name = "sqlite-${version}";
    version = "3.27.2+replication3";
    src = fetchFromGitHub {
      owner = "CanonicalLtd";
  # [...]
```

which is now [broken](https://hydra.nixos.org/log/c4vn91kdrllhlzshbvdv7rx3xpsxy2yj-sqlite-3.27.2+replication3.drv) because the patch won't apply (on 22.05).  But as it's much older version, I suspect it's accumulated vulnerabilities, too (same version on both nixpkgs branches).

**Comment by vcunat** [comment]
Perhaps /cc `sqlite` maintainer @np about the last comment.

**Pull request merged**